lib/middleware/index.js

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+/* jshint maxparams: 5 */
+
+var _ = require('underscore');
+
+/**
+ * A middleware is an object that implements any of the following methods:
+ *
+ * request(req, arg2, arg3, callback)
+ * response(req, err, res1, res2, callback)
+ *
+ * A middleware is free to alter req and update/replace any of arg2, arg3, err,
+ * res1 or res2. Multiple middlewares can be chained together forming a
+ * middleware stack. The callback is used to pass the control to the next
+ * middleware in chain. For example, a noop middleware can look like this:
+ *
+ * var noop = {
+ *     'request': function(req, arg2, arg3, callback) {
+ *         callback(arg2, arg3);
+ *     },
+ *     'response': function(req, err, res1, res2, callback) {
+ *         callback(err, res1, res2);
+ *     }
+ * };
+ *
+ * Given the middleware stack defined as:
+ *
+ * MiddlewareStack([m1, m2]).run(req, arg2, arg3, handler, handler_callback)
+ *
+ * The order in which the middlewares are applied in is: m1.request->
+ * m2.request-> handler-> m2.response-> m1.response-> handler_callback.
+ *
+ * The difference between applying a middleware stack to the server and the
+ * client is the logical direction a request and a response point to. For
+ * example, for a server it looks like this: server<-request, server->response;
+ * while for a client it's the reverse: client->request, client<-response.
+ */
+
+function MiddlewareStack(middlewares) {
+    this.middlewares = middlewares || [];
+}
+
+// XXX: adding nextTick to avoid blowing the call stack breaks a test
+MiddlewareStack.prototype.run = function run(req, arg2, arg3, handler, callback) {
+    var self = this;
+
+    var i = -1;
+    callRequestMiddleware(arg2, arg3);
+
+    function callRequestMiddleware(arg2, arg3) {
```

Comment:
Totally agree. I'm not familiar with the async lib, but thanks for the tip. I'll take a look and see if it can improve the readability. Either way, I think a comment is in place here.